## Ran the siderophile reproduction + re-enabled smoke test

The PR note said `cargo` wasn't installed here so the end-to-end `siderophile` reproduction and the re-enabled smoke test (#181) weren't run. I ran them with the rust toolchain (cargo 1.96.1) installed, applying this PR plus #181's re-enabled smoke test on top of `master`.

### Reproduction — confirms the fix works

`pytest test/test_smoke.py::TestSmoke::test_cargo --runintegration` (10m48s, resolves `trailofbits/siderophile@7bca0f5`):

- Resolution expands to **971 packages**, up from 129 on current `master` / 163 in the committed baseline. This matches the ~963 figure in #189 and verifies the fix restores "resolve all possible versions" for cargo.
- The new `CalledProcessError` handling fired as designed: 2 `cargo metadata` calls failed on fully-yanked transitive deps (`cloudflare-zlib-sys ^0.2.0`, `core2 ^0.4`) and were logged + skipped instead of aborting the whole run (the pre-fix behavior that #125's pattern addresses).
- `test/test_cargo.py`: 7 passed, including the new `test_resolve_from_source_only_marks_resolved_deps`.

### Smoke test — fails on one invariant, but it's baseline drift, not a regression

The only failure is:

```
AssertionError: resolution missing expected package(s) ['cargo:cc']
```

Root cause, verified from the resolved graph:

- In the 163-package baseline, `cc` was reachable **only** via `afl@0.10.1 → cc`.
- The full expansion now selects newer `afl` (`0.15.24`, `0.16.0`), which dropped the `cc` build-dependency (`afl` now deps `libc`, `rustc_version`, `xdg`).
- The remaining paths to `cc` (`flate2 → cloudflare-zlib-sys`, `bitstream-io → core2`) hit the fully-yanked crates above, so those subtrees legitimately error out.

So `cc` genuinely isn't in the current resolution. The baseline `siderophile.expected.json` predates both this fix and the current crates.io state — this is exactly the drift the "open question" here and #181 anticipate. The gap needs the semantics decision plus a re-baseline via `test/rebuild_expected_output.py`, not a code change. I did not rebuild the baseline, since that's the decision this draft defers.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

--- a/src/it_depends/cargo.py
+++ b/src/it_depends/cargo.py
@@ -120,6 +120,12 @@ def get_dependencies(
         for dep in package["dependencies"]:
             if dep["kind"] is not None:
                 continue
+            # Ignore optional (feature-gated) dependencies that are not explicitly required.
+            # A crate may declare a mutually exclusive set of optional deps, so forcing them all
+            # into the graph can describe a configuration that has no valid resolution. We only
+            # track dependencies cargo requires under the active feature set (see #189, #191).
+            if dep["optional"]:
+                continue
             if dep["name"] in dependencies:
                 dependencies[dep["name"]].semantic_version = dependencies[
                     dep["name"]
@@ -188,9 +194,9 @@ class CargoResolver(DependencyResolver):
                 else:
                     cache.add(package)
             # Only mark a dependency as resolved when `cargo metadata` actually resolved a
-            # matching package. Optional/feature-gated deps that the active feature set does
-            # not enable never appear in the resolved package set, so they must fall through
-            # to normal resolution rather than being marked resolved-to-nothing and pruned.
+            # matching package. A required dep that cargo failed to resolve (e.g. a subcrate
+            # whose `cargo metadata` errored) must fall through to normal resolution rather
+            # than being marked resolved-to-nothing and pruned.
             for package in packages:
                 for dep in package.dependencies:
                     if not cache.was_resolved(dep) and next(cache.match(dep), None) is not None:

--- a/test/test_cargo.py
+++ b/test/test_cargo.py
@@ -1,7 +1,9 @@
+import json
+
 import pytest
 
 from it_depends.cache import InMemoryPackageCache
-from it_depends.cargo import CargoResolver, _parse_workspace_member
+from it_depends.cargo import CargoResolver, _parse_workspace_member, get_dependencies
 from it_depends.models import Dependency, Package, SourcePackage
 from it_depends.repository import SourceRepository
 
@@ -29,18 +31,51 @@ def _dep(name: str) -> Dependency:
     return Dependency(package=name, source="cargo", semantic_version=CargoResolver.parse_spec("*"))
 
 
+def test_get_dependencies_ignores_optional_deps(monkeypatch: pytest.MonkeyPatch, tmp_path) -> None:  # noqa: ANN001
+    """Optional (feature-gated) deps are not tracked; only explicitly required deps survive.
+
+    Regression test for #189/#191: a crate may declare a mutually exclusive set of optional
+    deps, so forcing them all into the graph can describe a configuration with no valid
+    resolution. `get_dependencies` keeps only the deps cargo requires under the active feature
+    set, dropping optional and dev/build deps.
+    """
+    metadata = {
+        "packages": [
+            {
+                "name": "myapp",
+                "version": "0.1.0",
+                "dependencies": [
+                    {"name": "serde", "req": "^1", "kind": None, "optional": False},
+                    {"name": "rayon", "req": "^1", "kind": None, "optional": True},
+                    {"name": "criterion", "req": "^0.5", "kind": "dev", "optional": False},
+                ],
+            },
+        ],
+        "workspace_members": ["myapp 0.1.0 (path+file:///tmp/myapp)"],
+    }
+    monkeypatch.setattr(
+        "it_depends.cargo.subprocess.check_output",
+        lambda *args, **kwargs: json.dumps(metadata).encode(),  # noqa: ARG005
+    )
+
+    packages = list(get_dependencies(SourceRepository(path=tmp_path), cargo_path="cargo"))
+
+    assert len(packages) == 1
+    assert {dep.package for dep in packages[0].dependencies} == {"serde"}
+
+
 def test_resolve_from_source_only_marks_resolved_deps(monkeypatch: pytest.MonkeyPatch, tmp_path) -> None:  # noqa: ANN001
-    """Optional/feature-gated deps with no matching package must not be marked resolved.
+    """A required dep with no matching package must not be marked resolved.
 
     Regression test for #189: the old shortcut marked every declared dependency as resolved,
-    including optional deps that `cargo metadata` never adds to the resolved package set. Those
-    deps then matched nothing and their subtrees were pruned, collapsing "all possible versions"
-    to the locked tree.
+    including deps that `cargo metadata` never adds to the resolved package set (e.g. a subcrate
+    whose `cargo metadata` errored). Those deps then matched nothing and their subtrees were
+    pruned. Such deps must be left unresolved so normal resolution can retry them.
     """
     repo = SourceRepository(path=tmp_path)
 
-    # `libfoo` is actually resolved (a package is yielded); `libbar` and `rayon` are optional
-    # deps that cargo metadata never resolves, so no package is yielded for them.
+    # `libfoo` is actually resolved (a package is yielded); `libbar` and `libqux` are required
+    # deps that cargo failed to resolve, so no package is yielded for them.
     source_pkg = SourcePackage(
         name="myapp",
         version="0.1.0",
@@ -52,7 +87,7 @@ def test_resolve_from_source_only_marks_resolved_deps(monkeypatch: pytest.Monkey
         name="libfoo",
         version="1.0.0",
         source="cargo",
-        dependencies=[_dep("rayon")],
+        dependencies=[_dep("libqux")],
     )
 
     monkeypatch.setattr(CargoResolver, "can_resolve_from_source", lambda self, repo: True)  # noqa: ARG005
@@ -68,6 +103,6 @@ def test_resolve_from_source_only_marks_resolved_deps(monkeypatch: pytest.Monkey
     assert result is source_pkg
     # libfoo has a matching package in the cache -> resolved.
     assert cache.was_resolved(_dep("libfoo"))
-    # libbar and rayon resolved to nothing -> left unresolved so normal resolution can expand them.
+    # libbar and libqux resolved to nothing -> left unresolved so normal resolution can retry them.
     assert not cache.was_resolved(_dep("libbar"))
-    assert not cache.was_resolved(_dep("rayon"))
+    assert not cache.was_resolved(_dep("libqux"))

